### ISSUE TITLE
How do I delete an object in memory? A = nothing instead of A = 0

### ISSUE THREAD

**Issue by cossio** [JuliaLang/julia#24205]
Here is a quote from the documentation (https://docs.julialang.org/en/latest/manual/faq/)

> Julia does not have an analog of MATLAB's clear function; once a name is defined in a Julia session (technically, in module Main), it is always present.
>
>If memory usage is your concern, you can always replace objects with ones that consume less memory. For example, if A is a gigabyte-sized array that you no longer need, you can free the memory with A = 0. The memory will be released the next time the garbage collector runs; you can force this to happen with gc().

and see this discussion: https://discourse.julialang.org/t/why-julia-has-no-clear-variable/6541.

The proposed workaround, `A = 0`, has the problem that if I attempt to use the variable `A` at a later point, there will be no error, which can be unsafe when my intended behaviour wast to clear the variable to prevent its future use.

A better solution is to use `A = nothing` (as suggested by aaowens in the linked discusion), because most methods will fail with a Void and throw an error. I think the documentation should suggest `A = nothing` to "clear" variables instead of `A = 0`.

**Comment by samuela** [comment]
And what's the way to do this for functions? I'm getting 
```
julia> f = nothing
ERROR: invalid redefinition of constant f
Stacktrace:
 [1] top-level scope at none:0
```

**Comment by oscardssmith** [comment]
There isn't one.

**Comment by jhbdiehl** [comment]
I suppose there's still no solution to this?
I'm trying to work with a self-defined struct and have the same problem as samuela.

Pretty confusing to me that the official workflow for this is supposed to be "restart your repl", given that Julia 0.X seems to have had a clear() function and there are posts online from as far back as 2012 asking for this feature.

**Comment by cossio** [comment]
A possible workflow is to define your function within a module:

```
module A
f(x) = x
end
```

You can then later do:

```
module A
f = nothing
end
```

or rather just drop `f` altogether from the new definition of the module:

```
module A
end
```

**Comment by vergelli** [comment]
@cossio Would it let you use the "f" word?
I find this a bit workaroundish. Instead, Julia definitely needs a `del(x)` function or something similar.

**Comment by timholy** [comment]
Fixed in #33872. This is the better way to do it than `del` given that our compilation pipeline requires type-persistence (unlike certain other languages which don't have a compilation pipeline).

Demo:
```julia
julia> f = nothing

julia> f(x) = 1
ERROR: cannot define function f; it already has a value
Stacktrace:
 [1] top-level scope
   @ none:0
 [2] top-level scope
   @ REPL[2]:1

julia> module NewMain end
Main.NewMain

julia> using REPL

julia> REPL.activate(NewMain)

(Main.NewMain) julia> f(x) = 1
f (generic function with 1 method)

(Main.NewMain) julia> f(3)
1
```
